File: ios/CedarlingService.swift
import Foundation

internal final class CedarlingService {
  let queue = DispatchQueue(label: "com.cedarling.react-native.service", qos: .userInitiated)

  private var instance: Cedarling?
  private var lifecycleDisposed = false

  func initialize(bootstrapJson: String, archiveUri: String?) throws {
    try ensureUsable()
    try CedarlingJson.requireBootstrap(bootstrapJson)

    let replacement: Cedarling
    do {
      if let archiveUri {
        let archive = try CedarlingArchive.read(archiveUri)
        replacement = try Cedarling.loadFromJsonWithArchiveBytes(
          config: bootstrapJson,
          archiveBytes: archive
        )
      } else {
        replacement = try Cedarling.loadFromJson(config: bootstrapJson)
      }
    } catch let error as CedarlingModuleError {
      throw error
    } catch {
      throw CedarlingModuleError(.initialization, "Cedarling initialization failed")
    }

    let previous = instance
    instance = replacement
    release(previous)
  }

  func isInitialized() throws -> Bool {
    try ensureUsable()
    return instance != nil
  }

  func authorizeUnsigned(
    principalJson: String?,
    action: String,
    resourceJson: String,
    contextJson: String
  ) throws -> [String: Any] {
    try ensureUsable()
    let cedarling = try requireInstance()
    let validatedAction = try CedarlingJson.requireAction(action)
    try CedarlingJson.requireContext(contextJson)
    if let principalJson {
      try CedarlingJson.requireEntity(principalJson, label: "principal")
    }
    try CedarlingJson.requireEntity(resourceJson, label: "resource")

    do {
      let principal = try principalJson.map(entityFromJson)
      let resource = try entityFromJson(resourceJson)
      let context = JsonValue(contextJson)
      return try CedarlingResultMapper.map(
        cedarling.authorizeUnsigned(
          principal: principal,
          action: validatedAction,
          resource: resource,
          context: context
        )
      )
    } catch let error as CedarlingModuleError {
      throw error
    } catch {
      throw CedarlingModuleError(.authorization, "Cedarling authorization failed")
    }
  }

  func authorizeMultiIssuer(
    tokensJson: String,
    action: String,
    resourceJson: String,
    contextJson: String
  ) throws -> [String: Any] {
    try ensureUsable()
    let cedarling = try requireInstance()
    let validatedAction = try CedarlingJson.requireAction(action)
    try CedarlingJson.requireContext(contextJson)
    try CedarlingJson.requireEntity(resourceJson, label: "resource")
    let tokens = try CedarlingJson.parseTokens(tokensJson)

    do {
      let resource = try entityFromJson(resourceJson)
      let context = JsonValue(contextJson)
      return try CedarlingResultMapper.map(
        cedarling.authorizeMultiIssuer(
          tokens: tokens,
          action: validatedAction,
          resource: resource,
          context: context
        )
      )
    } catch let error as CedarlingModuleError {
      throw error
    } catch {
      throw CedarlingModuleError(.authorization, "Cedarling authorization failed")
    }
  }

  func getLogIds() throws -> [String] {
    try withInstance { $0.getLogIds() }
  }

  func getLogById(_ id: String) throws -> String {
    try withLogOperation {
      _ = try CedarlingJson.requireNonemptyString(id, label: "log id")
      return try $0.getLogById(id: id)
    }
  }

  func getLogsByRequestId(_ requestId: String) throws -> [String] {
    try withLogOperation {
      _ = try CedarlingJson.requireNonemptyString(requestId, label: "request id")
      return try $0.getLogsByRequestId(requestId: requestId)
    }
  }

  func getLogsByRequestIdAndTag(_ requestId: String, tag: String) throws -> [String] {
    try withLogOperation {
      _ = try CedarlingJson.requireNonemptyString(requestId, label: "request id")
      _ = try CedarlingJson.requireNonemptyString(tag, label: "log tag")
      return try $0.getLogsByRequestIdAndTag(requestId: requestId, tag: tag)
    }
  }

  func getLogsByTag(_ tag: String) throws -> [String] {
    try withLogOperation {
      _ = try CedarlingJson.requireNonemptyString(tag, label: "log tag")
      return try $0.getLogsByTag(tag: tag)
    }
  }

  func popLogs() throws -> [String] {
    try withLogOperation { try $0.popLogs() }
  }

  func pushDataContext(key: String, valueJson: String, ttlSeconds: Double?) throws {
    try withDataOperation {
      _ = try CedarlingJson.requireNonemptyString(key, label: "data context key")
      try CedarlingJson.requireJsonValue(valueJson, label: "data context value")
      try $0.pushDataCtx(
        key: key,
        value: valueJson,
        ttlSecs: try validateTtlSeconds(ttlSeconds)
      )
    }
  }

  func getDataContext(_ key: String) throws -> String? {
    try withDataOperation {
      _ = try CedarlingJson.requireNonemptyString(key, label: "data context key")
      return try $0.getDataCtx(key: key)
    }
  }

  func getDataContextEntry(_ key: String) throws -> [String: Any]? {
    try withDataOperation {
      _ = try CedarlingJson.requireNonemptyString(key, label: "data context key")
      return try $0.getDataEntryCtx(key: key).map(mapDataEntry)
    }
  }

  func removeDataContext(_ key: String) throws -> Bool {
    try withDataOperation {
      _ = try CedarlingJson.requireNonemptyString(key, label: "data context key")
      return try $0.removeDataCtx(key: key)
    }
  }

  func clearDataContext() throws {
    try withDataOperation { try $0.clearDataCtx() }
  }

  func listDataContext() throws -> [[String: Any]] {
    try withDataOperation { try $0.listDataCtx().map(mapDataEntry) }
  }

  func getDataContextStats() throws -> [String: Any] {
    try withDataOperation { mapDataStoreStats(try $0.getStatsCtx()) }
  }

  func isTrustedIssuerLoadedByName(_ name: String) throws -> Bool {
    try withInstance {
      _ = try CedarlingJson.requireNonemptyString(name, label: "trusted issuer name")
      return $0.isTrustedIssuerLoadedByName(issuerId: name)
    }
  }

  func isTrustedIssuerLoadedByIssuer(_ issuer: String) throws -> Bool {
    try withInstance {
      _ = try CedarlingJson.requireNonemptyString(issuer, label: "trusted issuer URL")
      return $0.isTrustedIssuerLoadedByIss(issClaim: issuer)
    }
  }

  func trustedIssuerSummary() throws -> [String: Any] {
    try withInstance {
      [
        "total": Double($0.totalIssuers()),
        "loaded": Double($0.loadedTrustedIssuersCount()),
        "loadedIds": $0.loadedTrustedIssuerIds(),
        "failedIds": $0.failedTrustedIssuerIds()
      ]
    }
  }

  func dispose() throws {
    try ensureUsable()
    let previous = instance
    instance = nil
    release(previous)
  }

  func nativeInfo() throws -> [String: Any] {
    try ensureUsable()
    return [
      "sdkVersion": CedarlingProvenance.sdkVersion,
      "cedarlingRevision": CedarlingProvenance.cedarlingRevision,
      "cedarlingCrateVersion": CedarlingProvenance.cedarlingCrateVersion,
      "uniffiVersion": CedarlingProvenance.uniffiVersion,
      "abis": CedarlingProvenance.supportedArchitectures
    ]
  }

  func disposeFromLifecycle() {
    queue.async { [self] in
      lifecycleDisposed = true
      let previous = instance
      instance = nil
      release(previous)
    }
  }

  private func ensureUsable() throws {
    if lifecycleDisposed {
      throw CedarlingModuleError(.alreadyDisposed, "Cedarling module has been destroyed")
    }
  }

  private func requireInstance() throws -> Cedarling {
    guard let instance else {
      throw CedarlingModuleError(.notInitialized, "Cedarling has not been initialized")
    }
    return instance
  }

  private func entityFromJson(_ json: String) throws -> EntityData {
    do {
      return try EntityData.fromJson(jsonString: json)
    } catch {
      throw CedarlingModuleError(.invalidInput, "entity conversion failed")
    }
  }

  private func release(_ cedarling: Cedarling?) {
    cedarling?.shutDown()
  }

  private func withInstance<T>(_ operation: (Cedarling) throws -> T) throws -> T {
    try ensureUsable()
    return try operation(requireInstance())
  }

  private func withLogOperation<T>(_ operation: (Cedarling) throws -> T) throws -> T {
    do {
      return try withInstance(operation)
    } catch let error as CedarlingModuleError {
      throw error
    } catch {
      throw CedarlingModuleError(.logging, "Cedarling log operation failed")
    }
  }

  private func withDataOperation<T>(_ operation: (Cedarling) throws -> T) throws -> T {
    do {
      return try withInstance(operation)
    } catch let error as CedarlingModuleError {
      throw error
    } catch {
      throw CedarlingModuleError(.dataContext, "Cedarling data-context operation failed")
    }
  }

  private func validateTtlSeconds(_ value: Double?) throws -> Int64? {
    guard let value else {
      return nil
    }
    guard value.isFinite,
          value >= 0,
          value.rounded(.towardZero) == value,
          value <= 9_007_199_254_740_991 else {
      throw CedarlingModuleError(
        .invalidInput,
        "ttlSeconds must be a non-negative safe integer"
      )
    }
    return Int64(value)
  }

  private func mapDataEntry(_ entry: DataEntry) -> [String: Any] {
    [
      "key": entry.key,
      "valueJson": entry.value,
      "dataType": entry.dataType,
      "createdAt": entry.createdAt,
      "expiresAt": entry.expiresAt,
      "accessCount": Double(entry.accessCount)
    ]
  }

  private func mapDataStoreStats(_ stats: DataStoreStats) -> [String: Any] {
    [
      "entryCount": Double(stats.entryCount),
      "maxEntries": Double(stats.maxEntries),
      "maxEntrySize": Double(stats.maxEntrySize),
      "metricsEnabled": stats.metricsEnabled,
      "totalSizeBytes": Double(stats.totalSizeBytes),
      "averageEntrySizeBytes": Double(stats.avgEntrySizeBytes),
      "capacityUsagePercent": stats.capacityUsagePercent,
      "memoryAlertThreshold": stats.memoryAlertThreshold,
      "memoryAlertTriggered": stats.memoryAlertTriggered
    ]
  }
}

File: android/src/main/java/expo/modules/cedarling/CedarlingService.kt
package expo.modules.cedarling

import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.jans.cedarling.CedarlingAndroid
import uniffi.cedarling_uniffi.Cedarling
import uniffi.cedarling_uniffi.DataEntry
import uniffi.cedarling_uniffi.DataStoreStats
import uniffi.cedarling_uniffi.EntityData
import uniffi.cedarling_uniffi.TokenInput

internal class CedarlingService(context: Context) {
  private val applicationContext = context.applicationContext
  private val mutex = Mutex()
  private val lifecycleScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

  @Volatile
  private var instance: Cedarling? = null

  @Volatile
  private var lifecycleDisposed = false

  suspend fun initialize(bootstrapJson: String, archiveUri: String?) =
    onIo {
      mutex.withLock {
        ensureUsable()
        CedarlingJson.requireBootstrap(bootstrapJson)
        CedarlingAndroid.ensureInitialized(applicationContext)
        val replacement = try {
          if (archiveUri == null) {
            Cedarling.loadFromJson(bootstrapJson)
          } else {
            val archive = CedarlingArchive.read(applicationContext.contentResolver, archiveUri)
            Cedarling.loadFromJsonWithArchiveBytes(bootstrapJson, archive)
          }
        } catch (error: CedarlingSdkException) {
          throw error
        } catch (error: Exception) {
          throw CedarlingSdkException(
            CedarlingErrorCode.INITIALIZATION,
            error.messageOrFallback("Cedarling initialization failed"),
            error
          )
        }

        val previous = instance
        instance = replacement
        previous?.let(::release)
      }
    }

  suspend fun isInitialized(): Boolean =
    onIo {
      mutex.withLock {
        ensureUsable()
        instance != null
      }
    }

  suspend fun authorizeUnsigned(
    principalJson: String?,
    action: String,
    resourceJson: String,
    contextJson: String
  ): Map<String, Any> =
    onIo {
      mutex.withLock {
        ensureUsable()
        val cedarling = requireInstance()
        CedarlingJson.requireAction(action)
        CedarlingJson.requireContext(contextJson)
        principalJson?.let { CedarlingJson.requireEntity(it, "principal") }
        CedarlingJson.requireEntity(resourceJson, "resource")

        val principal = principalJson?.let(::entityFromJson)
        val resource = entityFromJson(resourceJson)
        try {
          CedarlingResultMapper.map(
            cedarling.authorizeUnsigned(principal, action, resource, contextJson)
          )
        } catch (error: CedarlingSdkException) {
          throw error
        } catch (error: Exception) {
          throw CedarlingSdkException(
            CedarlingErrorCode.AUTHORIZATION,
            error.messageOrFallback("Cedarling authorization failed"),
            error
          )
        } finally {
          principal?.destroy()
          resource.destroy()
        }
      }
    }

  suspend fun authorizeMultiIssuer(
    tokensJson: String,
    action: String,
    resourceJson: String,
    contextJson: String
  ): Map<String, Any> =
    onIo {
      mutex.withLock {
        ensureUsable()
        val cedarling = requireInstance()
        CedarlingJson.requireAction(action)
        CedarlingJson.requireContext(contextJson)
        CedarlingJson.requireEntity(resourceJson, "resource")
        val tokens: List<TokenInput> = CedarlingJson.parseTokens(tokensJson)
        val resource = entityFromJson(resourceJson)
        try {
          CedarlingResultMapper.map(
            cedarling.authorizeMultiIssuer(tokens, action, resource, contextJson)
          )
        } catch (error: CedarlingSdkException) {
          throw error
        } catch (error: Exception) {
          throw CedarlingSdkException(
            CedarlingErrorCode.AUTHORIZATION,
            error.messageOrFallback("Cedarling authorization failed"),
            error
          )
        } finally {
          resource.destroy()
        }
      }
    }

  suspend fun getLogIds(): List<String> = withInstance { it.getLogIds() }

  suspend fun getLogById(id: String): String =
    withLogOperation {
      CedarlingJson.requireNonemptyString(id, "log id")
      it.getLogById(id)
    }

  suspend fun getLogsByRequestId(requestId: String): List<String> =
    withLogOperation {
      CedarlingJson.requireNonemptyString(requestId, "request id")
      it.getLogsByRequestId(requestId)
    }

  suspend fun getLogsByRequestIdAndTag(requestId: String, tag: String): List<String> =
    withLogOperation {
      CedarlingJson.requireNonemptyString(requestId, "request id")
      CedarlingJson.requireNonemptyString(tag, "log tag")
      it.getLogsByRequestIdAndTag(requestId, tag)
    }

  suspend fun getLogsByTag(tag: String): List<String> =
    withLogOperation {
      CedarlingJson.requireNonemptyString(tag, "log tag")
      it.getLogsByTag(tag)
    }

  suspend fun popLogs(): List<String> = withLogOperation { it.popLogs() }

  suspend fun pushDataContext(key: String, valueJson: String, ttlSeconds: Double?) =
    withDataOperation {
      CedarlingJson.requireNonemptyString(key, "data context key")
      CedarlingJson.requireJsonValue(valueJson, "data context value")
      it.pushDataCtx(key, valueJson, validateTtlSeconds(ttlSeconds))
    }

  suspend fun getDataContext(key: String): String? =
    withDataOperation {
      CedarlingJson.requireNonemptyString(key, "data context key")
      it.getDataCtx(key)
    }

  suspend fun getDataContextEntry(key: String): Map<String, Any>? =
    withDataOperation {
      CedarlingJson.requireNonemptyString(key, "data context key")
      it.getDataEntryCtx(key)?.let(::mapDataEntry)
    }

  suspend fun removeDataContext(key: String): Boolean =
    withDataOperation {
      CedarlingJson.requireNonemptyString(key, "data context key")
      it.removeDataCtx(key)
    }

  suspend fun clearDataContext() = withDataOperation { it.clearDataCtx() }

  suspend fun listDataContext(): List<Map<String, Any>> =
    withDataOperation { cedarling -> cedarling.listDataCtx().map(::mapDataEntry) }

  suspend fun getDataContextStats(): Map<String, Any> =
    withDataOperation { mapDataStoreStats(it.getStatsCtx()) }

  suspend fun isTrustedIssuerLoadedByName(name: String): Boolean =
    withInstance {
      CedarlingJson.requireNonemptyString(name, "trusted issuer name")
      it.isTrustedIssuerLoadedByName(name)
    }

  suspend fun isTrustedIssuerLoadedByIssuer(issuer: String): Boolean =
    withInstance {
      CedarlingJson.requireNonemptyString(issuer, "trusted issuer URL")
      it.isTrustedIssuerLoadedByIss(issuer)
    }

  suspend fun trustedIssuerSummary(): Map<String, Any> =
    withInstance {
      mapOf(
        "total" to it.totalIssuers().toDouble(),
        "loaded" to it.loadedTrustedIssuersCount().toDouble(),
        "loadedIds" to it.loadedTrustedIssuerIds(),
        "failedIds" to it.failedTrustedIssuerIds()
      )
    }

  suspend fun dispose() =
    onIo {
      mutex.withLock {
        ensureUsable()
        instance?.let(::release)
        instance = null
      }
    }

  fun disposeFromLifecycle() {
    lifecycleDisposed = true
    lifecycleScope.launch {
      mutex.withLock {
        instance?.let(::release)
        instance = null
      }
    }
  }

  fun nativeInfo(): Map<String, Any> {
    ensureUsable()
    return mapOf(
      "sdkVersion" to CedarlingProvenance.SDK_VERSION,
      "cedarlingRevision" to CedarlingProvenance.CEDARLING_REVISION,
      "cedarlingCrateVersion" to CedarlingProvenance.CEDARLING_CRATE_VERSION,
      "uniffiVersion" to CedarlingProvenance.UNIFFI_VERSION,
      "abis" to CedarlingProvenance.SUPPORTED_ABIS
    )
  }

  private fun ensureUsable() {
    if (lifecycleDisposed) {
      throw CedarlingSdkException(
        CedarlingErrorCode.ALREADY_DISPOSED,
        "Cedarling module has been destroyed"
      )
    }
  }

  private fun requireInstance(): Cedarling =
    instance ?: throw CedarlingSdkException(
      CedarlingErrorCode.NOT_INITIALIZED,
      "Cedarling has not been initialized"
    )

  private fun entityFromJson(json: String): EntityData =
    try {
      EntityData.fromJson(json)
    } catch (error: Exception) {
      throw CedarlingSdkException(
        CedarlingErrorCode.INVALID_INPUT,
        error.messageOrFallback("entity conversion failed"),
        error
      )
    }

  private fun release(cedarling: Cedarling) {
    try {
      cedarling.shutDown()
    } catch (_: Exception) {
      // Lifetime release must continue even if optional connection shutdown fails.
    } finally {
      cedarling.destroy()
    }
  }

  private suspend fun <T> withInstance(operation: (Cedarling) -> T): T =
    onIo {
      mutex.withLock {
        ensureUsable()
        operation(requireInstance())
      }
    }

  private suspend fun <T> withLogOperation(operation: (Cedarling) -> T): T =
    try {
      withInstance(operation)
    } catch (error: CedarlingSdkException) {
      throw error
    } catch (error: Exception) {
      throw CedarlingSdkException(
        CedarlingErrorCode.LOGGING,
        error.messageOrFallback("Cedarling log operation failed"),
        error
      )
    }

  private suspend fun <T> withDataOperation(operation: (Cedarling) -> T): T =
    try {
      withInstance(operation)
    } catch (error: CedarlingSdkException) {
      throw error
    } catch (error: Exception) {
      throw CedarlingSdkException(
        CedarlingErrorCode.DATA_CONTEXT,
        error.messageOrFallback("Cedarling data-context operation failed"),
        error
      )
    }

  private fun validateTtlSeconds(value: Double?): Long? {
    if (value == null) {
      return null
    }
    if (!value.isFinite() || value < 0 || value % 1.0 != 0.0 || value > 9_007_199_254_740_991.0) {
      throw CedarlingSdkException(
        CedarlingErrorCode.INVALID_INPUT,
        "ttlSeconds must be a non-negative safe integer"
      )
    }
    return value.toLong()
  }

  private fun mapDataEntry(entry: DataEntry): Map<String, Any> =
    mapOf(
      "key" to entry.key,
      "valueJson" to entry.value,
      "dataType" to entry.dataType,
      "createdAt" to entry.createdAt,
      "expiresAt" to entry.expiresAt,
      "accessCount" to entry.accessCount.toDouble()
    )

  private fun mapDataStoreStats(stats: DataStoreStats): Map<String, Any> =
    mapOf(
      "entryCount" to stats.entryCount.toDouble(),
      "maxEntries" to stats.maxEntries.toDouble(),
      "maxEntrySize" to stats.maxEntrySize.toDouble(),
      "metricsEnabled" to stats.metricsEnabled,
      "totalSizeBytes" to stats.totalSizeBytes.toDouble(),
      "averageEntrySizeBytes" to stats.avgEntrySizeBytes.toDouble(),
      "capacityUsagePercent" to stats.capacityUsagePercent,
      "memoryAlertThreshold" to stats.memoryAlertThreshold,
      "memoryAlertTriggered" to stats.memoryAlertTriggered
    )

  private suspend fun <T> onIo(block: suspend () -> T): T =
    withContext(Dispatchers.IO) { block() }
}
